src/Altinn.App.Core/Expressions/LayoutEvaluator.cs

```diff
@@ -0,0 +1,116 @@
+        {
+            HiddenFieldsForRemovalReucrs(state, hiddenModelBindings, nonHiddenModelBindings, childContext, hidden);
+        }
+
+        foreach (var (bindingName, binding) in context.Component.DataModelBindings)
+        {
+            if(bindingName == "group")
+            {
+                continue;
+            }
+
+            var indexed_binding = state.AddInidicies(binding, context);
+
+            if (hidden)
+            {
+                hiddenModelBindings.Add(indexed_binding);
+            }
+            else
+            {
+                nonHiddenModelBindings.Add(indexed_binding);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Remove fields that are only refrenced from hidden fields from the data object in the state.
+    /// </summary>
+    public static void RemoveHiddenData(LayoutEvaluatorState state)
+    {
+        var fields = GetHiddenFieldsForRemoval(state);
+        foreach (var field in fields)
+        {
+            state.RemoveDataField(field);
+        }
+    }
+
+    /// <summary>
+    /// Return a list of <see cref="ValidationIssue" /> for the given state and dataElementId
+    /// </summary>
+    public static IEnumerable<ValidationIssue> RunLayoutValidationsForRequired(LayoutEvaluatorState state, string dataElementId)
+    {
+        var validationIssues = new List<ValidationIssue>();
+
+        foreach (var context in state.GetComponentContexts())
+        {
+            RunLayoutValidationsForRequiredRecurs(validationIssues, state, dataElementId, context);
+        }
+
+        return validationIssues;
+    }
+
+    private static void RunLayoutValidationsForRequiredRecurs(List<ValidationIssue> validationIssues, LayoutEvaluatorState state, string dataElementId, ComponentContext context)
+    {
+        var hidden = ExpressionEvaluator.EvaluateBooleanExpression(state, context, "hidden", false);
+        if (!hidden)
+        {
+            foreach (var childContext in context.ChildContexts)
+            {
+                RunLayoutValidationsForRequiredRecurs(validationIssues, state, dataElementId, childContext);
+            }
+
+            var required = ExpressionEvaluator.EvaluateBooleanExpression(state, context, "required", false);
+            if (required && !hidden)
+            {
+                foreach (var (bindingName, binding) in context.Component.DataModelBindings)
+                {
+                    if (state.GetModelData(binding, context) is null)
+                    {
+                        validationIssues.Add(new ValidationIssue()
+                        {
+                            Severity = ValidationIssueSeverity.Error,
+                            InstanceId = state.GetInstanceContext("instanceId").ToString(),
+                            DataElementId = dataElementId,
+                            Field = state.AddInidicies(binding, context),
+                            Description = "TODO required",
```

Comment:
Yes, not sure what to write here. I have not figured out how validation really should work, and where this message is displayed.

I could possibly write something like 
```c#
Description = $"{context.Component.Id} is Required"
```
, but that might be a guid, that the user might not know anything about. Maybe I should use the `binding` (with indexes) instead, so that the error is tied to the datamodel, but that is already availible in the `Field` prop?